etcdserver/etcdhttp/client_security.go

```diff
@@ -0,0 +1,399 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdhttp
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/coreos/etcd/etcdserver"
+	"github.com/coreos/etcd/etcdserver/etcdhttp/httptypes"
+	"github.com/coreos/etcd/etcdserver/security"
+)
+
+type securityHandler struct {
+	sec         *security.Store
+	clusterInfo etcdserver.ClusterInfo
+}
+
+func hasWriteRootAccess(sec *security.Store, r *http.Request) bool {
+	if r.Method == "GET" || r.Method == "HEAD" {
+		return true
+	}
+	return hasRootAccess(sec, r)
+}
+
+func hasRootAccess(sec *security.Store, r *http.Request) bool {
+	if sec == nil {
+		// No store means no security avaliable, eg, tests.
+		return true
+	}
+	if !sec.SecurityEnabled() {
+		return true
+	}
+	username, password, ok := r.BasicAuth()
+	if !ok {
+		return false
+	}
+	if username != "root" {
+		return false
+	}
+	root, err := sec.GetUser("root")
+	if err != nil {
+		return false
+	}
+	return root.CheckPassword(password)
+}
+
+func hasKeyPrefixAccess(sec *security.Store, r *http.Request, key string) bool {
```

Comment:
shall we return a more specific error rather than a bool?

for example, now everything is 
```
{"message":"Insufficient credentials"}
```

i would prefer 

```
{"message":"bad password"}
```

```
{"message":"user does not exist"}
```

/cc @kelseyhightower  any idea?